### PR TITLE
Add normative changes from January 2023 plenary

### PR DESCRIPTION
**Pull request by rbuckton**
This adds the normative changes from the January 2023 plenary:

- Decorators are now also allowed before `export`
- Decorators are not allowed both before and after `export` or `export default`
- `context.access.get.call(obj)` -> `context.access.get(obj)`
- `context.access.set.call(obj, value)` -> `context.access.set(obj, value)`
- Adds `context.access.has(obj)`

Rather than using a production parameter like, this instead uses early errors to ensure a decorator list prior to `export` is only allowed for class declarations, and to ensure you do not have decorators both before and after `export`/`export default`.

### CODE REVIEW

**Comment by rbuckton** [comment]
@pzuraq, we should have these changes reviewed by the editors before merging.

File: spec.html
Comment (nicolo-ribaudo) [inline spec.html:28664]: I think it would be slightly easier to follow the grammar if it was defined like this:
```js
`export` LexicalDeclaration[~Yield, +Await]
`export` HoistableDeclaration[~Yield, +Await]
`export` ClassDeclaration[~Yield, +Await, +Decorators]
DecoratorList[~Yield, +Await]? `export` ClassDeclaration[~Yield, +Await, ~Decorators]

`export` `default` HoistableDeclaration[~Yield, +Await, +Default]
`export` `default` ClassDeclaration[~Yield, +Await, +Decorators, +Default]
DecoratorList[~Yield, +Await]? `export` `default` ClassDeclaration[~Yield, +Await, +Decorators, ~Default]
`export` `default` [lookahead &notin; { `function`, `async` [no LineTerminator here] `function`, `class`, `@` }] AssignmentExpression[+In, ~Yield, +Await] `;`
```

because the grammar rules are in a single place and you don't have to read both the grammar and the syntactic early errors. However, the end result is effectively the same so avoiding the new production parameter as you proposed is also ok :+1:

File: spec.html
Comment (nicolo-ribaudo) [inline spec.html]: Are `PrivateSet` and `Set` swapped (also before this PR)?

File: spec.html
Comment (rbuckton) [reply spec.html:28664]: I considered that, and that was what I'd implemented originally in https://github.com/tc39/proposal-decorators/pull/95/files. The early error seemed a little easier to reason over given that a `Decorators` parameter doesn't go very deep.

File: spec.html
Comment (rbuckton) [reply spec.html]: They are. I'll fix this, thanks.

File: spec.html
Comment (rbuckton) [reply spec.html:28664]: There's also a [bug in `esmeta`](https://github.com/es-meta/esmeta/issues/133) regarding multiple constraints on an RHS as would be needed here. That bug has since been fixed, but hasn't been released yet. I think there's still some work to go on the decorators PR before it will pass `esmeta` anyways, so it doesn't really matter too much either way.

**Comment by rbuckton** [comment]
@syg, @michaelficarra, @bakkot: Would one of you mind reviewing these changes? I understand the baseline spec in https://github.com/tc39/ecma262/pull/2417 (which this PR is a diff against) probably needs updating as well, but an editor's review on these individual changes would help so this can be merged into the main PR and unblock https://github.com/microsoft/TypeScript/pull/52582 and https://github.com/babel/babel/pull/15405.

File: spec.html
Comment (bakkot) [inline spec.html:24397]: The change to accept Symbols looks good, but just to confirm my understanding, that's a bugfix unrelated to the normative changes presented in committee, right?

(I have no problem with it going in with this PR, I just want to make sure I understand why it's here.)

File: spec.html
Comment (bakkot) [inline spec.html:29033]:
```suggestion
          1. If |DecoratorList| is present, then
```
I personally am neutral on whether the `?` should be included in general, but right now the prevailing style is that it is not. (See for example [ForLoopEvaluation](https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-runtime-semantics-forloopevaluation).)

File: spec.html
Comment (bakkot) [inline spec.html:29035]: Might be worth adding an assert here that |ClassDeclaration| does not have its own |DecoratorList| (and in the `export default` case as well, of course).

It's prevented by the early errors, but not by the mere structure of the nonterminals, so if you haven't noticed the early errors it seems like this is just discarding that |DecoratorList|.

File: spec.html
Comment (bakkot) [reply spec.html:28664]: I think I would prefer to have a new `UndecoratedClassDeclaration` nonterminal which derives what `ClassDeclaration` currently does in ecma262, and then have `ClassDeclaration` derive `DecoratorList? UndecoratedClassDeclaration`. Then you can use `UndecoratedClassDeclaration` here, no parameters or early errors necessary.

Still, that's an editorial change which can be made later. This change is correct as-is. Only reason to do it now is if it would cause TS or Babel to use a different representation in their ASTs; I know sometimes people like to have the AST shapes stick close to the spec (though this seems like a case where you would not want that).

**Review by bakkot — APPROVED** [review]
Couple tiny comments, but LGTM otherwise. Feel free to merge once the `?` and `assert` thing are addressed.

**Review by pzuraq — APPROVED** [review]

File: spec.html
Comment (rbuckton) [reply spec.html:24397]: Yes, its a bugfix. Without it, we'd try to do a `PrivateGet` using a symbol and not a Private Name, since _name_ is already declared to be a String, Symbol, or Private Name in the AO.

File: spec.html
Comment (rbuckton) [reply spec.html:29033]: I can change that, but there are other cases of `|DecoratorList?|` in the main PR we'll need to address at some point as well.

File: spec.html
Comment (rbuckton) [reply spec.html:24397]: > (I have no problem with it going in with this PR, I just want to make sure I understand why it's here.)

It's a bit of a drive-by bugfix since I needed to rewrite the algorithm steps for both `get` and `set` anyways and noticed the oversight.

File: spec.html
Comment (bakkot) [reply spec.html:29033]: Feel free to leave this as-is and address it as part of the main PR if you'd prefer, then. Either way.

File: spec.html
Comment (rbuckton) [reply spec.html:28664]: I don't know about Babel, but TypeScript won't introduce a new AST node to differentiate. We just store decorators as if they were modifiers now (a change we made in TS 4.9 to support decorators-after-`export` syntax), so `export`, `default`, `declare`, `abstract`, and decorators all end up in the same bucket. An `UndecoratedClassDeclaration` production feels more like a workaround for the grammar, much like cover grammars. I don't expect we would really tie any unique semantics to it, and it might even make handling `Function.prototype.toString` a bit more complicated.

File: spec.html
Comment (rbuckton) [reply spec.html:24397]: > Yes, its a bugfix. Without it, we'd try to do a `PrivateGet` using a symbol and not a Private Name, since _name_ is already declared to be a String, Symbol, or Private Name in the AO.

Correction: _name_ could already be a String, Symbol, or Private Name per the _name_ argument to the `CreateDecoratorContextObject` AO.

**Comment by rbuckton** [comment]
@pzuraq: I've addressed @bakkot's feedback. Can you merge this into your `decorators` branch so that it is applied to the main PR?

File: spec.html
Comment (bakkot) [reply spec.html:28664]: Eh, I disagree - using productions which precisely express what you actually want to accept is the normal use of a CFG, and grammar parameters and early errors are the workaround for when that's not practical.

Fair point about `Function.prototype.toString`, but I'd have to actually write it out to tell if it will actually be a problem.

**Comment by pzuraq** [comment]
Yes, thanks again for taking this on @rbuckton, super appreciated!

**Pull request merged**
